src/parsers/espresso/7.2/parser.js

```diff
@@ -0,0 +1,292 @@
+ * @param {Number} [c] - C parameter
+ * @returns {Number[]}
+ */
+function getCellConstants(celldm, a, b, c) {
+    let _a = celldm ? celldm[1] : a;
+    let _b = celldm ? celldm[2] * celldm[1] : b; // celldm[2] is b/a
+    let _c = celldm ? celldm[3] * celldm[1] : c; // celldm[3] is c/a
+    if (celldm) {
+        [_a, _b, _c] = [_a, _b, _c].map((x) => x * coefficients.BOHR_TO_ANGSTROM);
+    }
+    return [_a, _b, _c];
+}
+
+/**
+ * @summary Calculates cell angles from celldm(i) or cosAB, cosAC, cosBC parameters. Specific to Quantum Espresso.
+ * @param {Number[]} [celldm] - celldm(i) parameters
+ * @param {Number} [cosbc] - cosBC parameter
+ * @param {Number} [cosac] - cosAC parameter
+ * @param {Number} [cosab]   - cosAB parameter
+ * @returns {Number[]}
+ */
+function getCellAngles(celldm, cosbc, cosac, cosab) {
+    let alpha, beta, gamma;
+    if (cosbc) alpha = math.acos(cosbc);
+    if (cosac) beta = math.acos(cosac);
+    if (cosab) gamma = math.acos(cosab);
+
+    // Case for some of the cell types in QE docs
+    if (celldm && celldm[4]) {
+        gamma = math.acos(celldm[4]);
+    }
+
+    // Specific case for hexagonal cell in QE docs
+    if (celldm && celldm[4] && celldm[5] && celldm[6]) {
+        alpha = math.acos(celldm[4]);
+        beta = math.acos(celldm[5]);
+        gamma = math.acos(celldm[6]);
+    }
+
+    // Convert radians to degrees which are used in lattice definitions
+    [alpha, beta, gamma] = [alpha, beta, gamma].map((x) =>
+        x === undefined ? x : (x * 180) / math.PI,
+    );
+    return [alpha, beta, gamma];
+}
+
+/**
+ * @summary Returns cell config from ibrav and celldm(i) parameters
+ *
+ * QE docs: https://www.quantum-espresso.org/Doc/INPUT_PW.html#ibrav
+ * "If ibrav /= 0, specify EITHER [ celldm(1)-celldm(6) ]
+ *   OR [ A, B, C, cosAB, cosAC, cosBC ]
+ *   but NOT both. The lattice parameter "alat" is set to
+ *   alat = celldm(1) (in a.u.) or alat = A (in Angstrom);"
+ *
+ * @param {Object} system - The system parameters from &SYSTEM namelist
+ * @param {Number} system.ibrav - ibrav parameter
+ * @param {Number[]} [system.celldm] - celldm parameters
+ * @param {Number} [system.a] - A parameter in angstroms
+ * @param {Number} [system.b] - B parameter in angstroms
+ * @param {Number} [system.c] - C parameter in angstroms
+ * @param {Number} [system.cosab] - cosAB parameter
+ * @param {Number} [system.cosac] - cosAC parameter
+ * @param {Number} [system.cosbc] - cosBC parameter
+ * @returns {{vectors: Number[][], type: String}}
+ */
+function ibravToCellConfig(system) {
+    const { ibrav, celldm, a, b, c, cosab, cosac, cosbc } = system;
+    if (celldm && a) {
+        throw new Error("Both celldm and A are given");
+    } else if (!celldm && !a) {
+        throw new Error("Missing celldm(1)");
+    }
+
+    if (celldm) celldm.unshift(null); // Bump indices by 1 to make it easier to understand the logic. In QE input file celdm(1) array starts with 1, but parsed starting with 0.
```

Comment:
I see where you are coming from, but I don't think we should modify the array to use Fortran counting. It might be understandable in this function, but perhaps not so much in subsequent functions which use `celldm`. If you need the correspondence with `celldm(1)`, etc. I would suggest destructuring the array: `const [celldm1, celldm2, celldm3] = celldm;`

Reply:
I've shifted indices in functions where they are used with a comment added, as we discussed